fix(types)!: improve types in index.ts

AVaksman opened the pull request:
BREAKING CHANGE: cluster.setMetadata(): only node count is updatable on an existing cluster; getInstancesCallback/Response: dropped nextQuery property as it is deprecated for this method, exposed failedLocations property; instance.createCluster(): removed unsupported params serveNodes and defaultStorageType

- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Improve type annotations, some clean up and minor refactoring.

AVaksman commented on src/cluster.ts:
```diff
@@ -59,16 +59,21 @@ export type GetClustersCallback = (
   apiResponse?: google.bigtable.admin.v2.IListClustersResponse
 ) => void;
+export interface SetClusterMetadataOptions {
+  nodes: number;
```
Only `nodes` are updatable in cluster after it's creation

AVaksman commented on src/cluster.ts:
```diff
@@ -412,23 +414,9 @@ Please use the format 'my-cluster' or '${instance.name}/clusters/my-cluster'.`);
+      serveNodes: metadata.nodes,
     };
 
-    if (metadata.location) {
-      reqOpts.location = Cluster.getLocation_(
-        this.bigtable.projectId,
-        metadata.location
-      );
-    }
-
-    if (metadata.nodes) {
-      reqOpts.serveNodes = metadata.nodes;
-    }
-
-    if (metadata.storage) {
-      reqOpts.defaultStorageType = Cluster.getStorageType_(metadata.storage);
-    }
-
```
`storage` and `location` can not be updated on an existing cluster

stephenplusplus replied:
I haven't tested this or anything, but the docs say that the UpdateCluster rpc accepts a full "Cluster" object: https://cloud.google.com/bigtable/docs/reference/admin/rpc/google.bigtable.admin.v2#google.bigtable.admin.v2.BigtableInstanceAdmin.UpdateCluster-table

Will it return an error if certain fields are attempted to be updated?

AVaksman replied:
It won't return an error,
This change is done to eliminate a possibility of a confusion from the user side if attempted to change a filed that is not supported.

stephenplusplus replied:
I see, thanks. I'm not sure we should use the whitelist approach of only allowing "serveNodes", when the API could evolve someday to accept something else, and we'd be blocking it. Does the upstream team know their API docs state multiple unsupported properties can be sent during the UpdateCluster call, and maybe they would want to make it return an error from their end, or some other solution?

We could also throw from our library if we receive unsupported input, but in general, we're best to simply match what the official docs state is supported. As soon as we go off that path, things get shaky.

AVaksman replied:
> Does the upstream team know their API docs state multiple unsupported properties can be sent during the UpdateCluster call

While UpdateCluster rpc does accept a full "Cluster" object, it seems that upstream API docs do specify that among `Cluster` properties

- `name` is an _OutputOnly_ field
- `state` is an _OutputOnly_ field as well
- `location` is a _CreateOnly_ field
- `default_storage_type` is a _CreateOnly_ field as well
- leaving only `serveNodes ` field that can be updated

>  when the API could evolve someday to accept something else, and we'd be blocking it

If/when upstream API evolves to accept something else, that logic could be added as a new feature PR?

WDYT?

stephenplusplus replied:
> If/when upstream API evolves to accept something else, that logic could be added as a new feature PR?

There's usually a big lag before we notice.

If the upstream API silently drops properties, I would think our best bet is to do the same. In terms of code, let's pass them all through. In terms of types, maybe that's where you could just list "serveNodes"? At least then, if the user knows better and is trying to set a property we haven't added to the types yet, they could have their TypeScript compiler ignore that (I think?).

AVaksman replied:
Added the logic back, left the types with `nodes` only.

AVaksman replied:
@stephenplusplus WDYT?

AVaksman commented on src/index.ts:
```diff
@@ -49,14 +48,14 @@ const {grpc} = new gax.GrpcClient();
-    nextQuery?: {} | null,
-    response?: google.bigtable.admin.v2.IListInstancesResponse | null
+    result?: Instance[],
+    failedLocations?: string[],
```
BREAKING CHANGE
exposing [`failedLocations`](https://github.com/googleapis/googleapis/blob/0d7c3565dd942d0cc41b16e895046486ae29e44e/google/bigtable/admin/v2/bigtable_instance_admin.proto#L305) in the callback/response

stephenplusplus replied:
Let's make sure we add docs for this new response argument.

stephenplusplus replied:
Also, an example in the code block showing "if (failedLocations.length > 0) { // These locations contain instances which could not be retrieved. }"

AVaksman commented on src/instance.ts:
```diff
@@ -56,20 +58,15 @@ import {ServiceError} from 'google-gax';
 import {Bigtable} from '.';
 import {google} from '../protos/protos';
 
-export interface ClusterInfo {
-  id?: string;
-  location?: string;
-  serveNodes?: number;
-  nodes?: number;
-  storage?: string;
-  defaultStorageType?: number;
```
BREAKING CHANGE
Removing `serveNodes` and `defaultStorageType` as these are expected by GAPIC client, user facing parameters are `nodes` and `storage` respectively.

JustinBeckwith commented:
@AVaksman we prolly want to land this before shipping v3, right?  Can I trouble you to run through the open PRs and try to get these ready to go?

AVaksman commented:
BREAKING CHANGE: 
- cluster.setMetadata()
   - only node count is updatable on an existing cluster
   - removed `location` and `storage` from the params
- getInstancesCallback/Response
   - exposed `failedLocations` property that is also returned along with array of `Instnace`s
   - dropped `nextQuery` property as it is [deprecated](https://github.com/googleapis/googleapis/blob/2433bd50656264a2ef9f684bf646fb4d250d39ff/google/bigtable/admin/v2/bigtable_instance_admin.proto#L307-L308) for this method
- createInstance(id, options)
   - made `options` into a required param as it is not possible to create an instance without providing at least one `cluster` (part of `options` param)
- createCluster
   - removed duplicated, unsupported (by implementation), undocumented params `serveNodes` and `defaultStorageType` in favor of `nodes` and `location` respectfully

stephenplusplus commented on src/cluster.ts:
```diff
@@ -399,34 +401,36 @@ Please use the format 'my-cluster' or '${instance.name}/clusters/my-cluster'.`);
 
-    if (metadata.location) {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    if ((metadata as any).location) {
```
If the API actually drops these parameters, I think we can skip all of the logic where we worry about formatting them.

stephenplusplus commented on src/cluster.ts:
```diff
@@ -399,34 +401,36 @@ Please use the format 'my-cluster' or '${instance.name}/clusters/my-cluster'.`);
 
     const reqOpts: ICluster = {
       name: this.name,
+      serveNodes: metadata.nodes,
```
This is where I think we should pass all user-provided input straight through, as opposed to the whitelist approach where we only recognize certain parameters.

stephenplusplus replied:
(mapping `metadata.nodes` to `serveNodes` is still good)

AVaksman replied:
If applicable, user will need to pass the full
`location` as full path in the form of
`projects/<project>/locations/<zone_id>`

and a `defaultStorageType` (not `storage`) in upper case characters.

Done

stephenplusplus commented on test/cluster.ts:
```diff
@@ -374,65 +374,57 @@ describe('Bigtable/Cluster', () => {
     });
 
-    it('should respect the nodes option', done => {
+    it('should accept and pass user provided unput through', done => {
```
```suggestion
    it('should accept and pass user provided input through', done => {
```

AVaksman replied:
Fixed, thanks.

stephenplusplus commented on src/index.ts:
```diff
@@ -546,14 +544,9 @@ export class Bigtable {
    */
   createInstance(
     id: string,
-    optionsOrCallback?: InstanceOptions | CreateInstanceCallback,
```
Let's throw an error now that this is required (and add a test):

```ts
if (!options) {
  throw new Error('A configuration object is required to create an instance.');
}
```

AVaksman replied:
Done, thanks!

stephenplusplus commented on src/index.ts:
```diff
@@ -629,26 +624,21 @@ export class Bigtable {
+   *
+   *   if (data[1]) {
+   *     // These locations contain instances which could not be retrieved.
+   *     const failedLocations = data[1]
```
```suggestion
   *     const failedLocations = data[1];
```

AVaksman replied:
fixed, thanks!

stephenplusplus commented on src/index.ts:
```diff
@@ -547,6 +547,11 @@ export class Bigtable {
     options: InstanceOptions,
     callback?: CreateInstanceCallback
   ): void | Promise<CreateInstanceResponse> {
+    if (!options.clusters) {
```
Since we require an object, let's throw if that's missing, too. Otherwise, this would be a TypeError ("Cannot read property 'clusters' of undefined")

AVaksman replied:
Great point!
Done!

AVaksman commented:
@stephenplusplus before merging, does PR description look OK?

> BREAKING CHANGE: cluster.setMetadata(): only node count is updatable on an existing cluster; getInstancesCallback/Response: dropped nextQuery property as it is deprecated for this method, exposed failedLocations property; instance.createCluster(): removed unsupported params serveNodes and defaultStorageType

stephenplusplus commented:
It looks good to me!
